Allow string keys in `eval` utility

Raj-Parekh24 opened the pull request:
<!-- !! Thank your for opening a PR !! -->

### Motivation for these changes
Closes #227 

### Implementation details
 I have implemented a function that converts the dictionary with string keys to pytensor. The pytensor variable for the corresponding name is found using the function "get_var_by_name".

### Checklist
+ [x] Explain motivation and implementation 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues, preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes). Note that if they don't, we will [rewrite/rebase/squash the git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) before merging.
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇


## Major / Breaking Changes
- ...

## New features
- ... Its and enhancement of current functionality eval present in graph/basic.py under Variable class. Through this enhancement eval will accept string i.e. pytensor node name. 

## Bugfixes
- ...

## Documentation
- ...

## Maintenance
- ...

ricardoV94 commented on tests/graph/test_basic.py:
```diff
@@ -302,6 +302,14 @@ def test_eval(self):
+    def test_eval_with_strings(self):
+        assert self.w.eval({"x": 1.0, "y": 2.0}) == 6.0
+        assert self.w.eval({self.z: 3}) == 6.0
+        assert hasattr(self.w, "_fn_cache"), "variable must have cache after eval"
+        assert not hasattr(
+            pickle.loads(pickle.dumps(self.w)), "_fn_cache"
+        ), "temporary functions must not be serialized"
+
```
I don't think we need these checks. Instead more useful would be to include a mixed case like `.eval({"x": 1.0, y: 2.0})` and check the cases where the string does not match any variable name.

Finally we may want to add one test case with repeated variable names such as:
```python
x = pt.scalar()
x.name="x"
x = x + 1
x.name = "x"
y = pt.exp(x)
y.eval({"x": 0})
```
I think we should fail in those cases. Does `get_var_by_name` return all the instances with the name?

Raj-Parekh24 replied:
Hi @ricardoV94 ,

Thank you for the review.

`get_var_by_name` returns all instances of the same name. I am choosing the first declared variable i.e. last variable from the list to replace it with the name key.

ricardoV94 reviewed (commented):
This looks great @Raj-Parekh24 

I left some suggestions below

Raj-Parekh24 commented:
While running `pre-commit run --all` in the local machine I am getting a series of errors with the last message as `142/177 files pass as expected.`. But when I check its logs for basic.py there are no errors from my changes.

Is this expected behavior or I have broken something?

Raj-Parekh24 commented:
And in the last build, the test case failed for multiple environments is:
`FAILED tests/link/jax/test_scalar.py::test_erfinv - AssertionError: `

ricardoV94 commented:
> And in the last build, the test case failed for multiple environments is:
> `FAILED tests/link/jax/test_scalar.py::test_erfinv - AssertionError: `

Yeah needs to be investigated. Probably a new release of jax that behaves differently.

The tracking issue is https://github.com/pymc-devs/pytensor/issues/241

ricardoV94 commented:
Mypy doesn't work on python 3.11, issue here https://github.com/pymc-devs/pytensor/issues/240

Raj-Parekh24 commented:
Thanks for the help @ricardoV94 ,

I have updated the code and added warning when multiple variables are there for same name.

And also updated the test cases accordingly.

Please review it.

ricardoV94 commented on pytensor/graph/basic.py:
```diff
@@ -558,6 +558,46 @@ def get_parents(self):
             return [self.owner]
         return []
 
+    def convert_string_keys_to_pytensor_variables(self, inputs_to_values):
```
Let's make this an internal function of `eval`, no need to have it accessible to other functions in an outer scope

Raj-Parekh24 replied:
Ok then I will insert logic within the implementation part only.

ricardoV94 commented on pytensor/graph/basic.py:
```diff
@@ -558,6 +558,46 @@ def get_parents(self):
+                    raise Exception(f"{i} not found in graph")
+                else:
+                    if length_of_nodes_with_matching_names > 1:
+                        warnings.warn(
+                            f"Found {length_of_nodes_with_matching_names} pytensor variables with name {i} taking the first declared named variable for computation"
+                        )
```
It seems safer to just fail.

Raj-Parekh24 replied:
Instead of warning will throw Exception instead.

ricardoV94 commented on pytensor/graph/basic.py:
```diff
@@ -558,6 +558,46 @@ def get_parents(self):
+        Examples
+        --------
+
+        >>> import numpy as np
+        >>> import pytensor.tensor as at
+        >>> x = at.dscalar('x')
+        >>> y = at.dscalar('y')
+        >>> z = x + y
+        >>> np.allclose(z.eval({'x' : 3, 'y' : 1}), 4)
+        True
```
This does not show a use of the function. In any case we don't need a docstring with an example I think

Raj-Parekh24 replied:
Since we are making it internal to eval I also think that the doc-string will be no more required.

ricardoV94 commented on pytensor/graph/basic.py:
```diff
@@ -558,6 +558,46 @@ def get_parents(self):
             return [self.owner]
         return []
 
+    def convert_string_keys_to_pytensor_variables(self, inputs_to_values):
```
Given we are in `pytensor`, no need to mention it in the name. If you want you can specify the return type is `Dict[Variable, Variable]`

```suggestion
    def convert_string_keys_to_variables(self, inputs_to_values):
```

ricardoV94 commented:
@Raj-Parekh24 nice work. I left some comments above. Let me know what you think.

Raj-Parekh24 commented on tests/graph/test_basic.py:
```diff
@@ -302,6 +304,13 @@ def test_eval(self):
+        assert self.w.eval({self.z: 3}) == 6.0
+
+    def test_eval_with_strings_with_mulitple_same_name(self):
+        assert self.t.eval({"e": 1.0}) == 2.0
```
ok, instead of creating variables in setup_method will add the variables in test_function.

Raj-Parekh24 commented:
Hi @ricardoV94 ,

Thanks for the review and the suggestions, I have updated the implementation accordingly and pushed a new commit regarding it. Please review it and suggest the required changes.

Raj-Parekh24 commented:
Hey @ricardoV94 ,

I have updated the code bases on your suggestions. 

And I also want to work on issue #55, I got the requirements but still not able to understand how to implement it. Since I am new to the repo, can you please give me starting steps to work on it.

ricardoV94 commented:
Hi @Raj-Parekh24 sorry for the delay. I'll try to review your PR tomorrow. For that other PR I will give some more pointers there

Thanks for all the work!

Raj-Parekh24 commented:
Thanks for all the help and support @ricardoV94 .

ricardoV94 commented on tests/graph/test_basic.py:
```diff
@@ -302,6 +302,17 @@ def test_eval(self):
+        assert self.w.eval({"x": 1.0, self.y: 2.0}) == 6.0
+        assert self.w.eval({self.z: 3}) == 6.0
+
+    def test_eval_errors_having_mulitple_variables_same_name(self):
```
Should also add a test with zero variables matching since you have a specific error message for that case.

Raj-Parekh24 replied:
Thank you for your review @ricardoV94. Will add the corresponding test case.

ricardoV94 reviewed (commented):
Looks good. I left some suggestions below

Raj-Parekh24 commented on pytensor/graph/basic.py:
```diff
@@ -597,6 +597,30 @@ def eval(self, inputs_to_values=None):
+                        process_input_to_values[
+                            nodes_with_matching_names[
+                                length_of_nodes_with_matching_names - 1
+                            ]
```
Thank you for your review @ricardoV94. Will update the index value.

Raj-Parekh24 commented:
Thanks a lot for your review @ricardoV94,

I have added the code based on what you suggested. Please review it and kindly give your suggestions.

ricardoV94 reviewed (changes requested):
Functionality-wise this PR is ready! I just left some comments on code formatting. Let me know if you have any questions.

Raj-Parekh24 commented:
Thanks for the review @ricardoV94 ,

I have incorporated the suggested changes in the new commit, please review it.

ricardoV94 reviewed (approved):
Looks good, running the tests now!
